src/arithmetic/algebraic_expression.c

```diff
@@ -310,93 +209,301 @@ void AlgebraicExpression_PrependTerm(AlgebraicExpression *ae, GrB_Matrix m, bool
+
+        for(int i = 0; i < labelCount; i++) {
+            char *label = astEdge->labels[i];
+            Schema *s = GraphContext_GetSchema(gc, label, SCHEMA_EDGE);
+            if(!s) continue;
+            GrB_Matrix l = Graph_GetRelationMatrix(g, s->id);
+            GrB_Info info = GrB_eWiseAdd_Matrix_Semiring(m, NULL, NULL, Rg_structured_bool, m, l, NULL);
         }
+        mat = m;
+    }
+
+    op.operand = mat;
+    op.free = freeMatrix;
+    op.transpose = transpose;
+    return op;
+}
+
+AlgebraicExpression *AlgebraicExpression_Empty() {
+    return _AE_MUL(1);
+}
+
+AlgebraicExpression **AlgebraicExpression_From_QueryGraph(const QueryGraph *qg, const AST *ast, size_t *exp_count) {
+    /* Construct algebraic expression(s) from query graph 
+     * trying to take advantage of long multiplications with as few 
+     * transpose as possible we'll transform paths crossing the graph 
+     * "diameter", these are guarantee to be the longest, although
+     * there might be situations in which these are not the most optimal paths
+     * to explore.
+     * 
+     * Once a path been transformed it's removed from the graph and the process
+     * repeat itself. */
+
+    assert(qg);
+    
+    /* A graph with no edges implies an empty algebraic expression
+     * the reasoning behind this decission is that algebraic expression 
+     * represent graph traversals, no edges means no traversals. */
+    if(qg->edge_count == 0) {
+        *exp_count = 0;
+        return NULL;
+    }
 
-        if(dest->label) {
-            GrB_Matrix destMat = Node_GetMatrix(dest);
-            AlgebraicExpression_AppendTerm(exp, destMat, false, false);
+    Edge *e;
+    Node *src;
+    Node *dest;
+    AlgebraicExpression *exp;
+    AlgebraicExpressionOperand op;
+    AlgebraicExpression **exps = array_new(AlgebraicExpression*, 1);
+    QueryGraph *g = QueryGraph_Clone(qg);
+
+    // TODO: Have a single AST function which computes its referred entities.
+    TrieMap *ref_entities = NewTrieMap();
+    SetClause_ReferredEntities(ast->setNode, ref_entities);
+    WithClause_ReferredEntities(ast->withNode, ref_entities);
+    WhereClause_ReferredEntities(ast->whereNode, ref_entities);
+    ReturnClause_ReferredEntities(ast->returnNode, ref_entities);
+    CreateClause_ReferredEntities(ast->createNode, ref_entities);
+    DeleteClause_ReferredEntities(ast->deleteNode, ref_entities);
+    _referred_edge_ends(g, ref_entities);
+    _referred_variable_length_edges(g, ref_entities);
+
+    // As long as there are nodes to process.
+    while(g->node_count) {
+        // Get leaf nodes at the deepest level.
+        int level;
+        Node **leafs = _DeepestLevel(g, &level);
+        assert(array_len(leafs) > 0 && level >= 0);
+
+        // Get a path of length level.
+        Edge **path = DFS(leafs[0], level);
+        assert(array_len(path) == level);
+
+        // Construct expression.
+        exp = _AE_MUL(level * 3 - 1);  // (3) Node Edge Node.
```

Comment:
Shouldn't this be `level * 2 + 1`? Intermediate nodes are only counted once, so each level adds 1 edge and 1 node, plus 1 for the final destination. At level 1, this would only allocate 2.